cli/command/image/build.go

```diff
@@ -282,22 +298,41 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
 		progressOutput = &lastProgressOutput{output: progressOutput}
 	}
 
+	// if up to this point nothing has set the context then we must have have
+	// another way for sending it(streaming) and set the context to the Dockerfile
 	if dockerfileCtx != nil && buildCtx == nil {
 		buildCtx = dockerfileCtx
 	}
-	var s *session.Session
-	if isSessionSupported(dockerCli) {
-		sharedKey, err := getBuildSharedKey(contextDir)
-		if err != nil {
-			return errors.Wrap(err, "failed to get build shared key")
-		}
-		s, err = session.NewSession(filepath.Base(contextDir), sharedKey)
-		if err != nil {
-			return errors.Wrap(err, "failed to create session")
-		}
+
+	s, err := trySession(dockerCli, contextDir)
+	if err != nil {
+		return err
+	}
+
+	var body io.Reader
+	if buildCtx != nil && !options.stream {
+		body = progress.NewProgressReader(buildCtx, progressOutput, 0, "", "Sending build context to Docker daemon")
 	}
 
-	var body io.Reader = progress.NewProgressReader(buildCtx, progressOutput, 0, "", "Sending build context to Docker daemon")
+	// add context stream to the session
+	if options.stream && s != nil {
```

Comment:
I already moved some of it to the session file (and actually think it became less readable). This is synchronized with the build runner so needs to be in the same function. We could, of course, split things up like opts setting of context preparing but this is already large enough.